v3/lints/rfc/lint_key_usage_incorrect_length.go

```diff
@@ -0,0 +1,64 @@
+package rfc
+
+/*
+ * ZLint Copyright 2022 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import (
+	"fmt"
+	"math/big"
+
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/v3/lint"
+	"github.com/zmap/zlint/v3/util"
+)
+
+type keyUsageIncorrectLength struct{}
+
+func init() {
+	lint.RegisterLint(&lint.Lint{
+		Name:          "e_key_usage_incorrect_length",
+		Description:   "The key usage is a bit string with exactly nine possible flags",
+		Citation:      "RFC 5280: 4.2.1.3",
+		Source:        lint.RFC5280,
+		EffectiveDate: util.RFC5280Date,
+		Lint:          NewKeyUsageIncorrectLength,
+	})
+}
+
+func NewKeyUsageIncorrectLength() lint.LintInterface {
+	return &keyUsageIncorrectLength{}
+}
+
+func (l *keyUsageIncorrectLength) CheckApplies(c *x509.Certificate) bool {
+	return util.IsExtInCert(c, util.KeyUsageOID)
+}
+
+func (l *keyUsageIncorrectLength) Execute(c *x509.Certificate) *lint.LintResult {
+	keyUsage := util.GetExtFromCert(c, util.KeyUsageOID).Value
+	// Tag: keyUsage[0]
+	// Length: keyUsage[1]
+	// Unused: keyUsage[2]
+	// The actual key usage...
+	ku := keyUsage[3:]
+	// Any combination of the nine bit flags is legal from perspective
+	// of this lint (although requirements elsewhere may limit the combinations).
+	//
+	// As such, any value greater-than-or-equal-to 512 (2**9) is out of range of the possible
+	// values for a key usage bit string.
+	if big.NewInt(0).SetBytes(ku).Int64() > 0b111111111 {
```

Comment:
Suppose we have the keyUsage ASN1
```
0x30 0x03 0x07 0x80 0x80
````
This is
* `0x30`: bitString,
* `0x03`: 3 bytes (including the unused-indicator byte),
* `0x07`: 7 bits unused in the last byte,
* `0x80 0x80`: the bit string `10000000 10000000`, or rather `100000001` followed by 7 unused 0s.

This is a valid keyUsage value for the purposes of this lint: it sets flags 0 (digitalSignature) and 8 (decipherOnly).

However, this code doesn't cut off the unused bits before turning them into a bigint. It wouldn't run `big.NewInt(0).SetBytes(0b100000001)`, it would run `big.NewInt(0).SetBytes(0b1000000010000000)`. And that's [32896](https://go.dev/play/p/TCK6Y1-NMv6), significantly greater than 512. So I _think_ my hypothetical valid KU value would show up as invalid via this lint.

And even if I'm wrong (which I hope I am!) the fact that I'm not sure means this lint isn't obviously-correct.

I think that the clearer thing to do here would be to rely on cryptobyte to parse the bitstring value for us:
```go
keyUsageExt := cryptobyte.String(util.GetExtFromCert(c, util.KeyUsageOI).Value)
var keyUsageVal encoding_asn1.BitString
ok := keyUsageExt.ReadASN1BitString(&out)
if !ok {
		return &lint.LintResult{Status: lint.Error, Details: fmt.Sprintf("the key usage (%v) extension is not parseable.", ku)}
}
if out.BitLength > 9 {
		return &lint.LintResult{Status: lint.Error, Details: fmt.Sprintf("the key usage (%v) contains a value that is out of bounds of the range of possible KU values.", ku)}
}
return &lint.LintResult{Status: lint.Pass}
```
Or something like that.

Reply:
Thank you @aarongable!

So cryptobyte isn't truncating away the trailing bytes for us either, but I swapped over to using it because it is doing some extra validation which is always nice.

In order to address the issue, I've taken to simply bit shifting the `int64` representation value to the right `unused` number of times before comparing to 512.